src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTableBasedOnTwoVars.java

```diff
@@ -0,0 +1,118 @@
+package se.liu.ida.hefquin.engine.datastructures.impl;
+
+import org.apache.jena.graph.Node;
+import org.apache.jena.sparql.core.Var;
+import org.apache.jena.sparql.engine.binding.Binding;
+import se.liu.ida.hefquin.engine.data.SolutionMapping;
+import se.liu.ida.hefquin.engine.data.impl.SolutionMappingUtils;
+
+import java.util.*;
+
+public class SolutionMappingsHashTableBasedOnTwoVars extends SolutionMappingsHashTable{
+    protected final Map<Node, List<Node>> mapNN = new HashMap<>();
+    protected final Map<Node, List<SolutionMapping>> mapNS = new HashMap<>();
+    protected final Var joinVar1;
+    protected final Var joinVar2;
+
+    public SolutionMappingsHashTableBasedOnTwoVars( final Var joinVar1, final Var joinVar2 ) {
+        this.joinVar1 = joinVar1;
+        this.joinVar2 = joinVar2;
+    }
+
+    @Override
+    public boolean add( final SolutionMapping e ) {
+        final Binding solMapBinding = e.asJenaBinding();
+        final Node n1 = solMapBinding.get(joinVar1);
+        final Node n2 = solMapBinding.get(joinVar2);
+
+        if (n1 == null || n2 == null){
+            throw new IllegalArgumentException();
+        }
+        else{
+            List<Node> varList = mapNN.get(n1);
+            if( varList == null){
+                varList = new ArrayList<>();
+                mapNN.put(n1, varList);
+            }
+            varList.add(n2);
```

Comment:
`n2` may already be contained in `varList`. In this case, there is no need to insert it twice.